### PR TITLE
New woh: conditional-config

### PR DESCRIPTION
**Pull request by rute-santos**
### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] be against the correct branch (features can only go into develop)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

### CODE REVIEW

**Comment by gregorydlogan** [comment]
Without looking at the code:
- Needs a note in the release notes
- _Could_ be moved to 9.x, although I don't actually care if we do.

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandler.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      String condition = operation.getConfiguration(key);
+      if (StringUtils.isEmpty(condition)) {
+        continue;
+      }
+      boolean result = false;
+      try {
+        result = WorkflowConditionInterpreter.interpret(condition.trim());
+        logger.debug("Evaluate condition: {}, result: {}", condition, result);
+      } catch (IllegalArgumentException e) {
+        logger.warn("Invalid condition to evaluate: {}.", condition);
```
Comment (mtneug) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandler.java]: This should be an error.

```suggestion
        logger.error("Invalid condition to evaluate: {}.", condition);
```

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandler.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        throw new WorkflowOperationException(String.format("Invalid condition: %s", condition.trim()), e);
+      }
+      // If condition true, set variable and return
+      if (result) {
+        String valuePropName = VALUE_PREFIX + key.substring(CONDITION_PREFIX.length());
+        value = operation.getConfiguration(valuePropName);
+        if (StringUtils.isEmpty(value)) {
+          logger.warn("Condition {}  evaluated to true, but no value informed to set {} variable in {}",
```
Comment (mtneug) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandler.java]: This should be an error.
```suggestion
          logger.error("Condition {}  evaluated to true, but no value informed to set {} variable in {}",
```

File: modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandler.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    // If we got here, no condition was evaluated true so we use the no-match configuration if there
+    String noMatch = operation.getConfiguration(NO_MATCH);
+    if (StringUtils.isNotEmpty(NO_MATCH)) {
+      properties.put(wfConfigName, noMatch);
+      logger.debug("Configuration key '{}' of workflow {} is set to value '{}'", wfConfigName, id, noMatch);
+      return createResult(workflowInstance.getMediaPackage(), properties, Action.CONTINUE, 0);
+    }
+    return createResult(Action.CONTINUE);
```
Comment (mtneug) [inline modules/workflow-workflowoperation/src/main/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandler.java]: I guess one could argue both ways: the operation evaluated a bunch of expressions, so it did something, but really nothing has changed after this operation. So maybe `SKIP` would be more appropriate? Regardless, I'm fine with either one.

File: modules/workflow-workflowoperation/src/test/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    Assert.assertEquals("value no-match", result.getProperties().get(CONFIGURATION_VAR));
+  }
+
+  private WorkflowInstance replaceVars(WorkflowInstance wf, String key, String value) throws WorkflowParsingException {
+    HashMap<String, String> wfConfig = new HashMap<String, String>();
+    wfConfig.put(key, value);
+
+    // Replace operation configuration with variables form wf config
+    final Function<String, String> systemVariableGetter = k -> null;
+    String xml = WorkflowConditionInterpreter.replaceVariables(WorkflowParser.toXml(workflowInstance),
```
Comment (mtneug) [inline modules/workflow-workflowoperation/src/test/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandlerTest.java]: `workflowInstance` is used directly.
```suggestion
    String xml = WorkflowConditionInterpreter.replaceVariables(WorkflowParser.toXml(wf),
```

File: docs/guides/admin/docs/workflowoperationhandlers/conditional-config-woh.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+---------------
+
+|configuration key|description|example|
+|-----------------|-----------|-------|
+|configuration-name|Name of workflow configuration variable to be set|encoding-profile|
+|condition-**X**|Condition to be tested, same syntax as workflow conditions|(${presenter_work_resolution_x} &gt; 1600)|
+|value-**X**|Value to be assigned to configuration variable if condition-<X> is true|encoding-profile-x|
+|no-match|Value to be assigned to configuration variable if none of the conditions are true|encoding-profile-none|
+
+All _condition-**X**_ are sorted as strings and then evaluated in sequence e.g. condition-1, condition-2, etc.
```
Comment (mtneug) [inline docs/guides/admin/docs/workflowoperationhandlers/conditional-config-woh.md:32]: Could you add a description of what happens if multiple conditions are `true`? Conditions are evaluated in alphabetical order, which might be surprising (`condition-10` is evaluated before `condition-2`).

**Review by mtneug — CHANGES_REQUESTED** [review]
I still need to run a test workflow, but it looks good from what I see so far. Only some minor complaints.

File: docs/guides/admin/docs/workflowoperationhandlers/conditional-config-woh.md (hunk 5 above)
Comment (rute-santos) [reply docs/guides/admin/docs/workflowoperationhandlers/conditional-config-woh.md:32]: Will do. Yes, they are sorted alphabetically. The idea was that the part after the '-' could be anything, not only numbers, but I can change it to numbers if you think it is better. Please let me know.

File: modules/workflow-workflowoperation/src/test/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandlerTest.java (hunk 4 above)
Comment (rute-santos) [reply modules/workflow-workflowoperation/src/test/java/org/opencastproject/workflow/handler/workflow/ConditionalConfigWorkflowOperationHandlerTest.java]: Good catch!

File: docs/guides/admin/docs/workflowoperationhandlers/conditional-config-woh.md (hunk 5 above)
Comment (mtneug) [reply docs/guides/admin/docs/workflowoperationhandlers/conditional-config-woh.md:32]: No, I would also prefer any string. Just wanted to document the behaviour ;).

**Review by mtneug — APPROVED** [review]

**Comment by mtneug** [comment]
Works for me.

**Pull request merged**